### PR TITLE
Fix ENOSPC / EINVAL caused by chunk padding

### PR DESCRIPTION
**Pull request by jhermsmeier**
Connects to: https://github.com/resin-io/etcher/issues/1109

**Changes:**
- Adjusted `options.minimumChunkSize` to the lowest physical block size
- Updated `finalChunkSize` calculation to pick the smallest padding aligning to the minimum block size

### CODE REVIEW

**Comment by jviotti** [comment]
Can we add a test case with an image with the reported characteristics?

File: lib/write.js [hunk 1]
```diff
@@ -146,14 +146,7 @@ exports.usingStreaming = (deviceFileDescriptor, options = {}) => {
...
           if (chunk.length < options.chunkSize) {
-            const finalChunkSize = _.attempt(() => {
-              if (chunk.length > options.minimumChunkSize) {
-                return options.chunkSize;
-              }
-
-              return options.minimumChunkSize;
-            });
-
+            const finalChunkSize = Math.ceil(chunk.length / options.minimumChunkSize) * options.minimumChunkSize;
```
Comment (jviotti) [inline lib/write.js]: What happens if the chunk length is some very badly aligned number, like 127? If so, `finalChunkSize` will also be 127, which will for sure throw EINVAL. If the chunk length is smaller than the minimum chunk size, then we should pad with zeroes in order to get to that minimum chunk size.

Comment (jhermsmeier) [inline lib/write.js]: I'm not sure I follow.
If the chunk length were `127`, then above would yield `Math.ceil( 127 / 512 ) * 512)`, where the `Math.ceil()` would return `1` as `127 / 512 = ~0.248`, and the chunk would still be padded to `512` bytes.

Comment (jviotti) [inline lib/write.js]: Oh, sorry about that, it looks like I read it as if the ceil operation included the multiplication as well :)

**Comment by lurch** [comment]
Probably outside the scope of this PR, but tangentially related - maybe we could throw an error if the image we're given to write doesn't have a size in bytes exactly divisible by 512? Since AFAIK no drive has a blocksize smaller than 512 bytes, so a disk image that's not divisible by 512 is likely to be corrupt / malformed?

@jviotti Do we have any logged statistics that we can analyse to see if anybody ever _has_ tried writing an image that isn't a multiple of 512 bytes?

**Comment by jviotti** [comment]
I think we should still fix it (by padding) and continue. If you're interested in data, you can probably dig into Mixpanel (ask Apostolis for access), where we log every image size.

File: lib/write.js (hunk 1 above)
Comment (lurch) [inline lib/write.js]: LGTM - I guess this is basically rounding up the `finalChunksize` to the nearest multiple of 512 bytes?

**Comment by jhermsmeier** [comment]
> doesn't have a size in bytes exactly divisible by 512?

It doesn't matter whether the image is divisible by some number – the only reason the buffer about to be written needs to be multiple of the block size is the physical nature of block devices. You could starting writing in 8 KB blocks, switch to 64 KB blocks and then end with a 512 B block – but if you'd attempt to write a 768 B block or some other odd-number-of-bytes block, you'll get an `E_INVAL`, regardless of the source of the bytes.

> Since AFAIK no drive has a blocksize smaller than 512 bytes, so a disk image that's not divisible by 512 is likely to be corrupt / malformed?

That would mark images omitting trailing zero bytes, and probably many others we wouldn't think about as corrupt. I don't think that's a good heuristic for image health, especially since we have other means of verification available that are much stronger (checksums, bmaps).

**Comment by lurch** [comment]
> the only reason the buffer about to be written needs to be multiple of the block size is the physical nature of block devices.

Yeah, I'm aware of all that :)

> That would mark images omitting trailing zero bytes

Any idea if that actually happens / gets used?

> especially since we have other means of verification available that are much stronger (checksums, bmaps).

Yeah, but a simple file-size check is _much_ faster than either of those ;-)  (and doesn't need to be downloaded first in the streaming case)

If I get time next week I'll have a look into Mixpanel and see what the statistics tell us...

**Comment by jhermsmeier** [comment]
> Can we add a test case with an image with the reported characteristics?

Added an image which is an odd number of bytes, while also not being a power of 2 – meaning there's no way around padding the input.

**Comment by lurch** [comment]
An odd number which _is_ a power of 2 would be impressive :laughing:

**Comment by jviotti** [comment]
@jhermsmeier Do you have your NUCs already? Do you mind quickly testing this with a real device before merging, mainly on Windows? The CI servers interact with files on the same file-system, and I've seen times where the CI servers pass, but Windows throws things like EINVAL on external devices.

**Comment by jhermsmeier** [comment]
@jviotti I do have the NUCs already, but no SSDs for them yet. There's one high case variant NUC I can fit with a spare HDD I have here, until the storage arrives. Will give this a shot, and try it out before merging.

**Comment by jhermsmeier** [comment]
Tested this on one of the NUCs (Windows 10) now, it appears to work just fine.

**Comment by jviotti** [comment]
Amazing, let's merge! :D

**Pull request merged**